File: app/Sources/GunkApp/Extract/LicenseDetector.swift
import Foundation

struct DetectedLicense: Equatable, Sendable {
  let detected: String
  let warning: String?
}

final class LicenseDetector {
  private let fileManager: FileManager

  init(fileManager: FileManager = .default) {
    self.fileManager = fileManager
  }

  func detect(sourceRoot: URL) throws -> DetectedLicense {
    guard let licenseURL = try topLevelLicenseFile(in: sourceRoot) else {
      return DetectedLicense(detected: "unknown", warning: nil)
    }

    let contents = try String(contentsOf: licenseURL, encoding: .utf8)
    let detected = spdxIdentifier(for: contents)
    let warning = isRestrictive(detected)
      ? "Restrictive source license detected: \(detected). Reuse may require extra review."
      : nil

    return DetectedLicense(detected: detected, warning: warning)
  }

  private func topLevelLicenseFile(in sourceRoot: URL) throws -> URL? {
    let children = try fileManager.contentsOfDirectory(
      at: sourceRoot,
      includingPropertiesForKeys: [.isRegularFileKey],
      options: [.skipsHiddenFiles]
    )

    return children.first { url in
      let name = url.lastPathComponent.lowercased()
      return name == "license"
        || name.hasPrefix("license.")
        || name == "copying"
        || name.hasPrefix("copying.")
    }
  }

  private func spdxIdentifier(for contents: String) -> String {
    let uppercased = contents.uppercased()

    if uppercased.contains("GNU AFFERO GENERAL PUBLIC LICENSE") {
      return versionedGPL(prefix: "AGPL", contents: uppercased)
    }
    if uppercased.contains("GNU LESSER GENERAL PUBLIC LICENSE") {
      return versionedGPL(prefix: "LGPL", contents: uppercased)
    }
    if uppercased.contains("GNU GENERAL PUBLIC LICENSE") {
      return versionedGPL(prefix: "GPL", contents: uppercased)
    }
    if uppercased.contains("APACHE LICENSE") {
      return "Apache-2.0"
    }
    if uppercased.contains("MIT LICENSE") || uppercased.contains("PERMISSION IS HEREBY GRANTED") {
      return "MIT"
    }
    if uppercased.contains("BSD 3-CLAUSE") {
      return "BSD-3-Clause"
    }
    if uppercased.contains("BSD 2-CLAUSE") {
      return "BSD-2-Clause"
    }
    if uppercased.contains("ISC LICENSE") {
      return "ISC"
    }
    if uppercased.contains("MOZILLA PUBLIC LICENSE") {
      return "MPL-2.0"
    }

    return "unknown"
  }

  private func versionedGPL(prefix: String, contents: String) -> String {
    if contents.contains("VERSION 3") {
      return "\(prefix)-3.0-or-later"
    }
    if contents.contains("VERSION 2") {
      return "\(prefix)-2.0-or-later"
    }

    return prefix
  }

  private func isRestrictive(_ identifier: String) -> Bool {
    identifier.hasPrefix("GPL")
      || identifier.hasPrefix("AGPL")
      || identifier.hasPrefix("LGPL")
  }
}

File: engine/src/extract/licenseDetector.ts
// License detection. Ported from app/Sources/GunkApp/Extract/LicenseDetector.swift.

import { readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";

export interface DetectedLicense {
  detected: string;
  warning: string | null;
}

export class LicenseDetector {
  detect(sourceRoot: string): DetectedLicense {
    const licenseFile = this.topLevelLicenseFile(sourceRoot);
    if (!licenseFile) {
      return { detected: "unknown", warning: null };
    }
    let contents: string;
    try {
      contents = readFileSync(licenseFile, "utf-8");
    } catch {
      return { detected: "unknown", warning: null };
    }
    const detected = this.spdxIdentifier(contents);
    const warning = this.isRestrictive(detected)
      ? `Restrictive source license detected: ${detected}. Reuse may require extra review.`
      : null;
    return { detected, warning };
  }

  private topLevelLicenseFile(sourceRoot: string): string | null {
    let entries: string[];
    try {
      entries = readdirSync(sourceRoot, { withFileTypes: true })
        .filter((e) => e.isFile() && !e.name.startsWith("."))
        .map((e) => e.name);
    } catch {
      return null;
    }
    const match = entries.find((name) => {
      const lower = name.toLowerCase();
      return lower === "license" || lower.startsWith("license.") || lower === "copying" || lower.startsWith("copying.");
    });
    return match ? join(sourceRoot, match) : null;
  }

  private spdxIdentifier(contents: string): string {
    const upper = contents.toUpperCase();
    if (upper.includes("GNU AFFERO GENERAL PUBLIC LICENSE")) return this.versionedGPL("AGPL", upper);
    if (upper.includes("GNU LESSER GENERAL PUBLIC LICENSE")) return this.versionedGPL("LGPL", upper);
    if (upper.includes("GNU GENERAL PUBLIC LICENSE")) return this.versionedGPL("GPL", upper);
    if (upper.includes("APACHE LICENSE")) return "Apache-2.0";
    if (upper.includes("MIT LICENSE") || upper.includes("PERMISSION IS HEREBY GRANTED")) return "MIT";
    if (upper.includes("BSD 3-CLAUSE")) return "BSD-3-Clause";
    if (upper.includes("BSD 2-CLAUSE")) return "BSD-2-Clause";
    if (upper.includes("ISC LICENSE")) return "ISC";
    if (upper.includes("MOZILLA PUBLIC LICENSE")) return "MPL-2.0";
    return "unknown";
  }

  private versionedGPL(prefix: string, contents: string): string {
    if (contents.includes("VERSION 3")) return `${prefix}-3.0-or-later`;
    if (contents.includes("VERSION 2")) return `${prefix}-2.0-or-later`;
    return prefix;
  }

  private isRestrictive(identifier: string): boolean {
    return identifier.startsWith("GPL") || identifier.startsWith("AGPL") || identifier.startsWith("LGPL");
  }
}
